### PR TITLE
feat(plugins): Plugin API refactoring & expansion

### LINKED ISSUE #2126: Protractor.waitForAngular - add posibility to wait for jQuery.ajax requests
Our and other projects uses Angular and some other third-party libs. 
Those libs in common use jQuery.ajax - which activity doesn't tracked by waitForAngular (and internally by $browser.notifyWhenNoOutstandingRequests).

It will be good to waitForAngular internally checks for jQuery.active ==0 or provide an API to add custom additional checks.

Without this our test code become polluted by sleep or our custom waitForJquery (jQuery.active ==0) before each significant operation.

### PR DESCRIPTION
**Pull request by sjelin**
* Changed the plugin API so that it is more uniform (see docs/plugins.md)
* Updated existing plugins to the new API
* Added the `onPageLoad` and `onPageSync` entry points for plugins for modifying `browser.get`
* Added the `waitForPromise` and `waitForCondition` entry points for plugins for modifying `waitForAngular`
* Added tests

This closes #2126 and helps out @andresdominguez

[Review on Reviewable](https://reviewable.io/reviews/angular/protractor/2104)

### CODE REVIEW

File: lib/protractor.js [hunk 1]
```diff
@@ -440,12 +440,21 @@ Protractor.prototype.get = function(destination, opt_timeout) {
 
   destination = this.baseUrl.indexOf('file://') === 0 ?
     this.baseUrl + destination : url.resolve(this.baseUrl, destination);
+  this.driver.wait(function() {
```
Comment (juliemr) [inline lib/protractor.js]: Why does this need to be a wait? We could just do:
`destinationPromise = self.plugins_.mutatUrl(destination);`

and then below

`return destinationPromise.then(function(destination) {
  return this.driver.get(destination)....
`

File: lib/protractor.js [hunk 2]
```diff
@@ -475,6 +484,10 @@ Protractor.prototype.get = function(destination, opt_timeout) {
   }, timeout,
   'waiting for page to load for ' + timeout + 'ms');
 
+  this.driver.wait(function() {
```
Comment (juliemr) [inline lib/protractor.js]: This could be just:
```js
self.driver.controlFlow().execute(function() {
  return self.plugins_.onPageLoad();
});

File: lib/protractor.js [hunk 3]
```diff
@@ -507,9 +520,12 @@ Protractor.prototype.get = function(destination, opt_timeout) {
...
 
   return this.executeScript_(
-      'angular.resumeBootstrap(arguments[0]);',
-      msg('resume bootstrap'),
-      moduleNames);
+    'angular.resumeBootstrap(arguments[0]);',
+    msg('resume bootstrap'),
+    moduleNames
+  ).then(function() {
+      return self.plugins_.onPageSync();
```
Comment (juliemr) [inline lib/protractor.js]: nit: too much indentation here

File: docs/plugins.md [hunk 4]
```diff
@@ -85,64 +85,185 @@ the output of Protractor by returning a results object.
 Plugins are node modules which export an object with the following API:
```
Comment (sjelin) [inline docs/plugins.md]: @andresdominguez Could you look through this and see if these plugin expansions will be sufficient for your needs?

**Comment by sjelin** [comment]
@juliemr Should be good for you to review the whole change now.

File: lib/frameworks/jasmine.js [hunk 5]
```diff
@@ -31,13 +31,15 @@ exports.run = function(runner, specs) {
   };
   RunnerReporter.prototype.reportSpecResults = function(spec) {
     var specInfo = {
-      name: spec.suite.getFullName(),
```
Comment (sjelin) [inline lib/frameworks/jasmine.js]: The changes in this file are just bug-fixes.

File: docs/plugins.md (hunk 4 above)
Comment (juliemr) [inline docs/plugins.md]: Ping @andresdominguez :)

File: docs/plugins.md [hunk 6]
```diff
@@ -85,64 +85,224 @@ the output of Protractor by returning a results object.
...
  *
- * @param {Object} config The plugin configuration object.
+ * @this {Object} bound to module.exports
+ *
+ * @throws {*} If this function throws an error, it is outputted to the console
  *
- * @return Return values are ignored.
+ * @return {q.Promise=} Can return a promise, in which case protractor will wait
+ *     for the promise to resolve before continuing.  If the promise is
+ *     rejected, a failed assertion is added to the test results.
```
Comment (juliemr) [inline docs/plugins.md]: Has this changed? I think `postResults` cannot change the test results.

File: docs/plugins.md [hunk 7]
```diff
@@ -85,64 +85,224 @@ the output of Protractor by returning a results object.
...
+ * @this {Object} bound to module.exports
+ *
+ * @throws {*} If this function throws an error, a failed assertion is added to
+ *     the test results and the original URL is used
+ *
+ * @return {string|q.Promise<string>} The modified URL or a promise which
+ *     resolves to the modified URL.  If a rejected promise is returned, a
+ *     failed assertion is added to the test results.
+ */
+exports.rewriteURL = function() {};
```
Comment (juliemr) [inline docs/plugins.md]: Wondering what the use case is for `rewriteURL` to return a promise?

File: docs/plugins.md [hunk 8]
```diff
@@ -85,64 +85,224 @@ the output of Protractor by returning a results object.
...
+ *
+ * @throws {*} If this function throws an error, a failed assertion is added to
+ *     the test results.
+ *
+ * @return {q.Promise=} Can return a promise, in which case protractor will wait
+ *     for the promise to resolve before continuing.  If the promise is
+ *     rejected, a failed assertion is added to the test results.
  */
-exports.postTest = function(config, passed, testInfo) {};
+exports.onPageSync = function() {};
```
Comment (juliemr) [inline docs/plugins.md]: Let's call this `onPageStable` to try to avoid overloading the `sync` term more. That also corresponds with the `whenStable` call we make to Angular Testability.

File: docs/plugins.md [hunk 9]
```diff
@@ -85,64 +85,224 @@ the output of Protractor by returning a results object.
...
+ * @throws {*} If this function throws an error, a failed assertion is added to
+ *     the test results.
+ *
+ * @return {q.Promise<boolean>|boolean} If truthy, Protractor will continue onto
+ *     the next command.  If falsy, webdriver will continuously re-run this
+ *     function until it is truthy.  If a rejected promise is returned, a failed
+ *     assertion is added to the test results, and protractor will continue onto
+ *     the next command.
+ */
+exports.waitForCondition = function() {};
```
Comment (juliemr) [inline docs/plugins.md]: Can we merge `waitForPromise` and `waitForCondition` into one method? Like how `webdriver.wait` can deal with both?

File: docs/plugins.md [hunk 10]
```diff
@@ -85,64 +85,224 @@ the output of Protractor by returning a results object.
...
+ *
+ * @type {Object}
+ */
+exports.config;
+
+/**
+ * Adds a failed assertion to the test's results.
+ *
+ * @param {string} message The error message for the failed assertion
+ * @param {Object=} options Some optional extra information about the assertion:
```
Comment (juliemr) [inline docs/plugins.md]: This should be
@param {specName: string, stackTrace: string}

File: lib/plugins.js [hunk 11]
```diff
@@ -70,87 +102,145 @@ Plugins.printPluginResults = function(specResults) {
...
-      return ret;
     });
+
+    return q.all(promises);
   };
 }
 
-/**
- * Sets up plugins before tests are run.
+/*
```
Comment (juliemr) [inline lib/plugins.js]: nit: /**

File: lib/plugins.js [hunk 12]
```diff
@@ -70,87 +102,145 @@ Plugins.printPluginResults = function(specResults) {
...
-    var promises = [];
-    for (var i = 0; i < this.pluginConfs.length; ++i) {
-      var pluginConf = this.pluginConfs[i];
-      var pluginObj = this.pluginObjs[i];
-      names.push(pluginObj.name || pluginConf.name);
-      promises.push(
-          (pluginObj[funName] || noop).apply(
-              pluginObj[funName],
-              [pluginConf].concat([].slice.call(arguments))));
+/*
```
Comment (juliemr) [inline lib/plugins.js]: nit: /**

File: lib/plugins.js [hunk 13]
```diff
@@ -70,87 +102,145 @@ Plugins.printPluginResults = function(specResults) {
...
-            log_('Plugin "' + names[i] + '" returned a malformed object');
-            continue; // Just ignore this result
-          }
-
-          // Output test results
-          if (pluginResult.specResults) {
-            log.puts('Plugin: ' + names[i] + ' (' + funName + ')');
-            Plugins.printPluginResults(pluginResult.specResults);
-          }
+/*
```
Comment (juliemr) [inline lib/plugins.js]: nit: /**

File: lib/plugins.js [hunk 14]
```diff
@@ -70,87 +102,145 @@ Plugins.printPluginResults = function(specResults) {
...
- *
- * @return {q.Promise} A promise which resolves when the plugins have all been
- *     torn down.
+ * @see docs/plugins.md#writing-plugins for information on these functions
  */
+Plugins.prototype.setup = pluginFunFactory('setup');
+Plugins.prototype.teardown = pluginFunFactory('teardown');
+Plugins.prototype.postResults = pluginFunFactory('postResults');
 Plugins.prototype.postTest = pluginFunFactory('postTest');
+Plugins.prototype.rewriteURL = pluginMutatorFunFactory('rewriteURL');
```
Comment (juliemr) [inline lib/plugins.js]: Knee-jerk reaction: I dislike the special infrastructure for the mutator, it seems more complicated. Does the URL rewriter belong in plugins? Can that be done in a better way?

File: spec/plugins/waitForAngularConf.js [hunk 15]
```diff
@@ -0,0 +1,52 @@
+var env = require('../environment.js'),
+    q = require('q');
+
+// A small suite to make sure the basic functionality of plugins work
+// Tests the (potential) edge case of exactly one plugin being used
```
Comment (juliemr) [inline spec/plugins/waitForAngularConf.js]: Update this comment?

**Comment by juliemr** [comment]
Reviewed! I have some concerns about the url rewriter complicating things, but overall I think this cleans stuff up a ton and is great.

File: docs/plugins.md (hunk 7 above)
Comment (sjelin) [inline docs/plugins.md]: Donno.  Didn't really take additional work, and it seemed consistent with everything else, so I did it.

File: docs/plugins.md (hunk 6 above)
Comment (sjelin) [inline docs/plugins.md]: Good catch.  However, protractor really will wait on a promise if it's returned.  And if the promise is rejected, an error will be logged to the console now

**Comment by andresdominguez** [comment]
It looks really good. I think it covers all of our requirements.

**Comment by sjelin** [comment]
Finally got around to this.

I decided against merging `waitForPromise` and `waitForCondition` because I couldn't find a way to handle the case of a promise resolving to `undefined` in a clean way.

I kept the `rewriteURL` implementation the same because mutator hooks in plugins will probably happen again at some point in the future.

**Comment by sjelin** [comment]
@juliemr Any new comments?

File: docs/plugins.md [hunk 16]
```diff
@@ -85,64 +85,227 @@ the output of Protractor by returning a results object.
...
+ *
+ * @return {string|q.Promise<string>} The modified URL or a promise which
+ *     resolves to the modified URL.  If a rejected promise is returned, a
+ *     failed assertion is added to the test results.
+ */
+exports.rewriteURL = function() {};
+
+/**
+ * This is called inside browser.get() directly after the page loads, and before
+ * angular has had time to bootstrap/synchronize.
```
Comment (juliemr) [inline docs/plugins.md]: Let's change this to:

'before angular bootstraps'.

Since it's not a matter of it having time, since we control when bootstrapping happens.

**Comment by juliemr** [comment]
As discussed offline - let's do this without the mutator stuff for now. Otherwise, LGTM.

**Comment by sjelin** [comment]
Closed with f2a11a7

**Comment by mfulton26** [comment]
Can waitForPromise and waitForCondition be made to work for non-AngularJS apps? Currently setting ignoreSynchronization to true causes custom waitForPromise/waitForCondition functions to not be called at all.
